Enable diag for bool Tensors

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47455 Enable diag for bool Tensors**
* #47454 Add bool tensor support for where

Differential Revision: [D24772483](https://our.internmc.facebook.com/intern/diff/D24772483)

---
### Description
Fixes #30923
This now works:

```
x = torch.tensor([True, False, True], dtype=torch.bool, device='cpu')
res = torch.diag(x)

x = torch.tensor([True, False, True], dtype=torch.bool, device='cuda')
res = torch.diag(x)
```

Also updated diag test to include bools and include checks for both cpu and gpu devices

Test:
pytest test/test_torch.py -k test_diag -vs

H-Huang committed:
Enable diag for bool Tensors

[ghstack-poisoned]

H-Huang committed:
Update on "Enable diag for bool Tensors"




[ghstack-poisoned]

izdeby commented on aten/src/ATen/native/cuda/TriangularOps.cu:
```diff
@@ -229,7 +229,7 @@ Tensor& apply_diag(Tensor& result, const Tensor& self, int64_t dimension) {
 
 Tensor& diag_cuda_out(Tensor& result, const Tensor& self, int64_t dimension) {
-  AT_DISPATCH_ALL_TYPES_AND(ScalarType::Half, self.scalar_type(), "diag_cuda", [&] {
+  AT_DISPATCH_ALL_TYPES_AND2(ScalarType::Half, at::ScalarType::Bool, self.scalar_type(), "diag_cuda", [&] {
```
nit: `at::` is not needed

H-Huang committed:
Update on "Enable diag for bool Tensors"


Differential Revision: [D24772483](https://our.internmc.facebook.com/intern/diff/D24772483)

---
### Description
Fixes #30923
This now works:

```
x = torch.tensor([True, False, True], dtype=torch.bool, device='cpu')
res = torch.diag(x)

x = torch.tensor([True, False, True], dtype=torch.bool, device='cuda')
res = torch.diag(x)
```

Also updated diag test to include bools and include checks for both cpu and gpu devices

Test:
pytest test/test_torch.py -k test_diag -vs

[ghstack-poisoned]
